File: 86.分隔链表.cpp
/*
 * @lc app=leetcode.cn id=86 lang=cpp
 *
 * [86] 分隔链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        auto n1 = new ListNode(0);
        auto n2 = new ListNode(0);
        auto tmp1 = n1;
        auto tmp2 = n2;
        while (head != nullptr) {
            if (head->val < x) {
                tmp1->next = head;
                tmp1 = tmp1->next;
                head = head->next;
                tmp1->next = nullptr;
            } else {
                tmp2->next = head;
                tmp2 = tmp2->next;
                head = head->next;
                tmp2->next = nullptr;
            }
        }
        tmp1->next = n2->next;
        return n1->next;
    }
};
// @lc code=end

File: 86.分隔链表.py
#
# @lc app=leetcode.cn id=86 lang=python
#
# [86] 分隔链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        tmp1, tmp2 = ListNode(0), ListNode(0)
        n1, n2 = tmp1, tmp2
        while head:
            if head.val < x:
                n1.next = head
                head = head.next
                n1 = n1.next
                n1.next = None
            else:
                n2.next = head
                head = head.next
                n2 = n2.next
                n2.next = None
            
        n1.next = tmp2.next
        return tmp1.next
# @lc code=end

